src/ILLink.RoslynAnalyzer/DynamicallyAccessedMembersAnalyzer.cs

```diff
@@ -53,33 +55,60 @@ public override void Initialize (AnalysisContext context)
 					}
 				}
 			});
-			// TODO: fix reporting for generic type substitutions. This should happen not only for method invocations,
-			// but for any reference to an instantiated method or type.
-			context.RegisterOperationAction (context => {
-				var invocationOperation = (IInvocationOperation) context.Operation;
-				ProcessInvocationOperation (context, invocationOperation);
-			}, OperationKind.Invocation);
+			context.RegisterSyntaxNodeAction (context => {
+				ProcessGenericParameters (context);
+			}, SyntaxKind.GenericName);
 		}
 
-		static void ProcessInvocationOperation (OperationAnalysisContext context, IInvocationOperation invocationOperation)
+		static void ProcessGenericParameters (SyntaxNodeAnalysisContext context)
 		{
-			if (context.ContainingSymbol.HasAttribute (RequiresUnreferencedCodeAnalyzer.RequiresUnreferencedCodeAttribute))
+			if (context.ContainingSymbol is not null && context.ContainingSymbol.HasAttribute (RequiresUnreferencedCodeAnalyzer.RequiresUnreferencedCodeAttribute)) {
 				return;
+			}
 
-			ProcessTypeArguments (context, invocationOperation);
-		}
+			ImmutableArray<ITypeParameterSymbol> typeParams = default;
+			ImmutableArray<ITypeSymbol> typeArgs = default;
+			if (context.SemanticModel.GetTypeInfo (context.Node).ConvertedType is INamedTypeSymbol type) {
+				// INamedTypeSymbol inside nameof, commonly used in [ExpectedWarning] can generate diagnostics
+				// Walking the node heirarchy to check if INamedTypeSymbol is inside a nameof
+				var node = context.Node;
+				while (node != null) {
+					if (node is InvocationExpressionSyntax invocationExpression && invocationExpression.Expression is IdentifierNameSyntax ident1) {
+						if (ident1.Identifier.ValueText.Equals ("nameof"))
+							return;
+					}
+					node = node.Parent;
+				}
+				typeParams = type.TypeParameters;
+				typeArgs = type.TypeArguments;
+			}
 
-		static void ProcessTypeArguments (OperationAnalysisContext context, IInvocationOperation invocationOperation)
-		{
-			var targetMethod = invocationOperation.TargetMethod;
-			if (targetMethod.HasAttribute (RequiresUnreferencedCodeAnalyzer.RequiresUnreferencedCodeAttribute))
-				return;
+			if (context.SemanticModel.GetSymbolInfo (context.Node, context.CancellationToken).Symbol is IMethodSymbol targetMethod) {
+				// Check if the generic parameters have any constraints
+				var node = context.Node;
+				while (node != null) {
+					if (node is MethodDeclarationSyntax parentMethod) {
+						if (parentMethod.ConstraintClauses.Count > 0) {
+							// This is safe, the linker would have marked the default .ctor already
+							return;
```

Comment:
I don't understand this:
* If there are constraints on the generic method, those should be effectively converted into annotations (or at least that's basically how linker treats this). For example `void Method<T>() where T : new {}` basically means that T should behave as if it has `DynamicallyAccessedMemberTypes.PublicParameterlessConstructors`. That said, Roslyn would have already checked this and failed if the instantiation doesn't match this, so we don't need to do anything here. On the other hand the T could have additional annotations of its own, in which case we still need to check the instantiation.
* Why is this done only for methods and not types?

Reply:
- Made the constraints on type parameters more granular by looking only for new(), struct and unmanaged.
- I couldn't find a test scenario in the GenericParameterDF tests but will look closely when enabling more tests. Roslyn seems to be keeping the constraint information on the Method, which required walking up the tree to get and is only handling the generic methods for now

Reply:
Moved constraints outside of this to [GenericParameterValue ](https://github.com/dotnet/linker/blob/8485bfcfbc55ab2b7007e3d39903e502fee82727/src/ILLink.RoslynAnalyzer/TrimAnalysis/GenericParameterValue.cs#L22)as @vitek-karas suggested